consistent_web_main/consistent_hash/simple_hash.go

```diff
@@ -0,0 +1,78 @@
+package consistent_hash
+
+import (
+	"sort"
+	"sync"
+	"time"
+)
+
+type SimpleHash struct {
+	orderedKeys    []string
+	nodeMap        map[string]ServerNode
+	sizeInclRepls  uint32
+	mux            sync.RWMutex
+}
+
+func NewSimpleHash(nodeMap map[string]ServerNode) *SimpleHash {
+	var size uint32 = 0
+	orderedKeys := make([]string, 0)
+
+	for ip, node := range nodeMap  {
+		size += (uint32) (node.Replicas)
+		orderedKeys = append(orderedKeys, ip)
+	}
+	h := &SimpleHash{
+		orderedKeys:                 orderedKeys,
+		nodeMap:                     nodeMap,
+		sizeInclRepls:               size,
+	}
+
+	return h
+}
+
+func (h *SimpleHash) InsertNode(ip_address string, replica_count int) {
+	h.mux.Lock()
+	defer h.mux.Unlock()
+	if entry, ok := h.nodeMap[ip_address]; ok {
+		h.sizeInclRepls -= (uint32) (entry.Replicas)
+		entry.Replicas = replica_count
+		h.sizeInclRepls += (uint32) (entry.Replicas)
+
+		// Note that this changes the distribution for other keys after the ip address
+		// this is fine because regardless the other keys will be changed
+		h.nodeMap[ip_address] = entry
+	} else {
+		timestamp := time.Now().Add(60 * time.Second)
+		h.nodeMap[ip_address] = ServerNode{IP: ip_address, Timestamp: timestamp, Replicas: replica_count}
+		h.orderedKeys = append(h.orderedKeys, ip_address)
+		h.sizeInclRepls += (uint32) (replica_count)
+	}
+}
+
+func (h *SimpleHash) DeleteNode(ip string) {
+	h.mux.Lock()
+	defer h.mux.Unlock()
+	h.sizeInclRepls -= (uint32) (h.nodeMap[ip].Replicas)
+	index := sort.Search(len(h.orderedKeys), func(i int) bool {
+		return h.orderedKeys[i] >= ip
+	})
+
+	if index < len(h.orderedKeys) && h.orderedKeys[index] == ip {
+		h.orderedKeys = append(h.orderedKeys[:index], h.orderedKeys[index+1:]...)
+	}
+	delete(h.nodeMap, ip)
+}
+
+func (h *SimpleHash) ValueLookup(value string) string {
+	h.mux.RLock()
+	defer h.mux.RUnlock()
+	replica_idx := getTrieKey(value) % h.sizeInclRepls
+	var tempSize uint32 = 0
+	for _, ip := range h.orderedKeys {
```

Comment:
(Comment, no need to change code here)
I think this is very inefficient because it is O(n) for value lookup. It would have been better to push the O(n) to the uncommon case (insert/remove nodes) and kept value lookup as O(1). This could have been done by simply storing all vnode replicas in orderedKeys array, so it would have looked like [ip1, ip1, ip1, ip2, ip3, ip3, ip3, ip3, ip4, ip5, ip5, ...]

That being said, since this is not the main focus of our project and we are also short of time, _we can let it be like it is right now_.